use round instead of mod for sinpi/cospi reduction

simonbyrne opened the pull request:
Now that `round` corresponds to native instruction on most processors, it is significantly faster to use it for the range reduction, as opposed to `mod` (which calls out to a libm routine, typically using x87 instructions).

I see a ~20% speedup.

simonbyrne committed:
avoid use of mod in sinpi/cospi for floating point types

ViralBShah commented:
@simonbyrne At the rate you are going, we may no longer need openlibm!

pao commented on base/special/trig.jl:
```diff
@@ -96,14 +96,21 @@ mulpi_ext(x::Float32) = DoubleFloat32(pi*Float64(x))
 mulpi_ext(x::Real) = pi*x # Fallback
 
-function sinpi(x::Real)
+function sinpix(x::Real)
```
This looks unintentional?

simonbyrne replied:
Good catch! Thanks.

StefanKarpinski commented:
Any reason not to merge?

tkelman commented:
> Any reason not to merge?

Test failures on all 5 architectures, only one of which was spurious?

tkelman commented on base/special/trig.jl:
```diff
@@ -100,10 +100,17 @@ function sinpi(x::Real)
     end
 
-    rx = copysign(rem(x,2),x)
+    hx = 0.5*x
```
would it be more type stable if this were `1//2 * x` or `x / 2` ?

simonbyrne replied:
Yep, that seems to be the issue. Good catch.

simonbyrne commented:
Okay, hopefully that should be fixed now.

tkelman commented:
Tests okay. Same performance conclusions?

simonbyrne commented:
Yes, still seeing 20% speed up or so.

The main downsides I can foresee:
* the "round" instruction is SSE4: I don't know how it will go on an earlier machine (or, if we care that much about performance on such machines).
* it is going to break completely if the user changes the rounding mode (due to the dependence of `round` on rounding mode), at least until LLVM gets around to exporting a `roundeven` intrinsic.

simonbyrne commented:
Actually, it just occurred to me that some of those branches are redundant: I might be able to make some more changes.

nalimilan commented:
@simonbyrne Couldn't we check for `JULIA_CPU_TARGET` to choose the most efficient instruction when building the sysimage?

The issue with rounding modes is a bit more frightening...

simonbyrne commented:
We could, but I don't have access to a pre-SSE4 machine for profiling. I suspect the overhead wouldn't be huge in any case.

As for the rounding mode: I feel that so many things behave oddly when you change the rounding mode  that we should just have a big "here be dragons" warning.

pao commented:
> the "round" instruction is SSE4: I don't know how it will go on an earlier machine

Or ARM targets? (It looks like there's at least a NEON instruction, but that's single-precision only.)

simonbyrne commented:
Actually, we can do this via addition/subtraction, which would be much more portable, and seems to have comparable performance.

simonbyrne commented:
Okay, updated.

I also changed `mulpi_ext`, as I found that zeroing trailing bits is much faster than using Veltkamp splitting (see https://gist.github.com/simonbyrne/62f25608aaf831de5996).

JeffBezanson commented:
@simonbyrne merge if and when you think this is ready.

simonbyrne commented:
Great, done.
